cpp/src/transform/one_hot_encode.cu

```diff
@@ -0,0 +1,145 @@
+#include <rmm/exec_policy.hpp>
+
+#include <algorithm>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+template <typename InputType, bool has_nulls>
+struct one_hot_encode_functor {
+  one_hot_encode_functor(column_device_view input, column_device_view category)
+    : _equality_comparator{input, category}, _input_size{input.size()}
+  {
+  }
+
+  bool __device__ operator()(size_type i)
+  {
+    size_type element_index  = i % _input_size;
+    size_type category_index = i / _input_size;
+    return _equality_comparator.template operator()<InputType>(element_index, category_index);
+  }
+
+ private:
+  element_equality_comparator<has_nulls> _equality_comparator;
+  size_type _input_size;
+};
+
+}  // anonymous namespace
+
+template <bool has_nulls>
+struct one_hot_encode_launcher {
+  template <typename InputType, CUDF_ENABLE_IF(is_equality_comparable<InputType, InputType>())>
+  std::pair<std::unique_ptr<column>, table_view> operator()(column_view const& input_column,
+                                                            column_view const& categories,
+                                                            rmm::cuda_stream_view stream,
+                                                            rmm::mr::device_memory_resource* mr)
+  {
+    auto total_size    = input_column.size() * categories.size();
+    auto all_encodings = make_numeric_column(
+      data_type{type_id::BOOL8}, total_size, mask_state::UNALLOCATED, stream, mr);
+
+    auto d_input_column    = column_device_view::create(input_column, stream);
+    auto d_category_column = column_device_view::create(categories, stream);
+    one_hot_encode_functor<InputType, has_nulls> one_hot_encoding_compute_f(*d_input_column,
+                                                                            *d_category_column);
+
+    thrust::transform(rmm::exec_policy(stream),
+                      thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(total_size),
+                      all_encodings->mutable_view().begin<bool>(),
+                      one_hot_encoding_compute_f);
+
+    auto split_iter = make_counting_transform_iterator(
+      1, [width = input_column.size()](auto i) { return i * width; });
+    std::vector<size_type> split_indices(split_iter, split_iter + categories.size() - 1);
+
+    // TODO: use detail interface, gh9226
+    auto views = cudf::split(all_encodings->view(), split_indices);
+    table_view encodings_view{views};
+
+    return std::make_pair(std::move(all_encodings), encodings_view);
+  }
+
+  template <typename InputType,
+            typename... Arg,
+            CUDF_ENABLE_IF(not is_equality_comparable<InputType, InputType>())>
+  std::pair<std::unique_ptr<column>, table_view> operator()(Arg&&...)
+  {
+    CUDF_FAIL("Cannot encode column type without well-defined equality operator.");
+  }
+};
+
+std::pair<std::unique_ptr<column>, table_view> one_hot_encode(column_view const& input_column,
```

Comment:
Declare this API in detail/transform.hpp and add a default value for the stream parameter.